src/IECoreScene/ShaderNetwork.cpp

```diff
@@ -332,33 +411,62 @@ class ShaderNetwork::Implementation
 			return true;
 		}
 
+
 		void hash( IECore::MurmurHash &h ) const
 		{
-			std::unique_lock<std::mutex> lock( m_hashMutex );
-			if( m_hashDirty )
+			update();
+			h.append( m_hash );
+		}
+
+		void hashSubstitutions( const CompoundObject *attributes, MurmurHash &h ) const
+		{
+			update();
+			for( const auto &a : m_neededSubstitutions )
 			{
-				m_hash = MurmurHash();
-				for( const auto &node : m_nodes )
+				const StringData *sourceAttribute = attributes->member<StringData>( a );
+				if( sourceAttribute )
 				{
-					m_hash.append( node.handle );
-					node.shader->hash( m_hash );
+					h.append( sourceAttribute->readable() );
+				}
+				else
+				{
+					// Need to append placeholder since which attributes are found matters
+					h.append( 0 );
+				}
+			}
+		}
 
-					for( const auto &connection : node.inputConnections )
+		void applySubstitutions( const CompoundObject *attributes )
+		{
+			update();
+
+			for( const auto &nodeAndParms : m_parmsNeedingSubstitution )
+			{
+				ShaderPtr s ( m_nodes.find( nodeAndParms.first )->shader->copy() );
+				for( const auto &parm : nodeAndParms.second )
+				{
+					StringData *targetParm = runTimeCast< StringData >( s->parameters()[parm].get() );
+					if( targetParm )
 					{
-						m_hash.append( connection.source.shader );
-						m_hash.append( connection.source.name );
-						m_hash.append( connection.destination.name );
+						targetParm->writable() = stringApplySubstitutions( targetParm->readable(), attributes );
+						continue;
 					}
-				}
 
-				m_hash.append( m_output.shader );
-				m_hash.append( m_output.name );
+					StringVectorData *targetParmVector = runTimeCast< StringVectorData >( s->parameters()[parm].get() );
+					if( targetParmVector )
+					{
+						std::vector<std::string> &stringVector = targetParmVector->writable();
+						for( unsigned int i = 0; i < stringVector.size(); i++ )
+						{
+							stringVector[i] = stringApplySubstitutions( stringVector[i], attributes );
+						}
+						continue;
+					}
 
-				m_hashDirty = false;
+					throw Exception( "ShaderNetwork::applySubstitutions : parameter need substitution couldn't be fond - was the network somehow modified without being dirtied?" );
```

Comment:
"need substitution couldn't be fond" -> "needing substitution couldn't be found"